Switch Optimizer to std::map

JacobSzwejbka opened the pull request:
Summary: Switch to map api which is directly compatible with TrainingModule

Reviewed By: davidlin54

Differential Revision: D62453507

dvorjackz commented on extension/training/test/training_loop_test.cpp:
```diff
@@ -30,73 +31,52 @@ using exec_aten::Tensor;
 
-  // Check that the data has changed.
-  ASSERT_NE(param_data[0].data_ptr<float>()[0], orig_data);
+  // // Check that the data has changed.
```
```suggestion
  // Check that the data has changed.
```

jackzhxng reviewed (approved):
Approving sans the sgd logic since David looked at it

JacobSzwejbka committed:
Switch Optimizer to std::map (#5230)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/5230

Switch to map api which is directly compatible with TrainingModule

Update the simple end 2 end test to use TrainingModule as well.

Reviewed By: davidlin54

Differential Revision: D62453507
